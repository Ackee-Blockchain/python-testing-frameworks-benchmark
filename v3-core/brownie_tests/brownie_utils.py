from contextlib import contextmanager

import brownie


@contextmanager
def brownie_reverts_fix(
    message=None,
):
    try:
        with brownie.reverts(message):
            yield
    except ValueError:
        pass
    except AssertionError as e:
        assert e.args[0] == f"Unexpected revert string 'None'\n"
    except Exception as e:
        assert isinstance(e, ValueError), f"Expected ValueError, got {e}"
