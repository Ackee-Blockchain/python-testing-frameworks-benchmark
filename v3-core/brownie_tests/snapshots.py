import json
from pathlib import Path


def match_snapshot(value, file_path, ident):
    filename = Path(__file__).parent / "__snapshots__" / f"{Path(file_path).name}.snap"
    with open(filename, "r") as f:
        expected = json.load(f)
        assert value == int(expected[ident])
