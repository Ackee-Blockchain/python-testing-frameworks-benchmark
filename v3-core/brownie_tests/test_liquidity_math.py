import brownie
import pytest
from brownie import LiquidityMathTest, accounts

from brownie_tests.snapshots import match_snapshot


@pytest.fixture(scope="function")
def liquidity_math():
    return LiquidityMathTest.deploy({"from": accounts[0]})


class TestLiquidityMath:
    def test_add_zero(self, liquidity_math):
        """
        1 + 0
        """
        assert liquidity_math.addDelta(1, 0) == 1

    def test_add_negative(self, liquidity_math):
        """
        1 + -1
        """
        assert liquidity_math.addDelta(1, -1) == 0

    def test_add_positive(self, liquidity_math):
        """
        1 + 1
        """
        assert liquidity_math.addDelta(1, 1) == 2

    def test_overflow(self, liquidity_math):
        """
        2**128-15 + 15 overflows
        """
        with brownie.reverts("LA"):
            liquidity_math.addDelta(2**128 - 15, 15)

    def test_underflow(self, liquidity_math):
        """
        0 + -1 underflows
        """
        with brownie.reverts("LS"):
            liquidity_math.addDelta(0, -1)

    def test_underflow_with_values(self, liquidity_math):
        """
        3 + -4 underflows
        """
        with brownie.reverts("LS"):
            liquidity_math.addDelta(3, -4)

    def test_gas_add(self, liquidity_math):
        """
        gas add
        """
        match_snapshot(
            liquidity_math.getGasCostOfAddDelta(15, 4),
            __file__,
            "addDelta_gas_add",
        )

    def test_gas_sub(self, liquidity_math):
        """
        gas sub
        """
        match_snapshot(
            liquidity_math.getGasCostOfAddDelta(15, -4),
            __file__,
            "addDelta_gas_sub",
        )
