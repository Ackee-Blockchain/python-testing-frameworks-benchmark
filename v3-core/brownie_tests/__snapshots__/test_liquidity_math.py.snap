{
    "addDelta_gas_add": "162",
    "addDelta_gas_sub": "176"
}